### PR TITLE
Consider aggregate bases when checking if an InitListExpr is constant

### PR DESCRIPTION
**Pull request by rnk**
This code was correct as written prior to C++17, which allowed bases to appear in the initializer list.

This was observable by creating non-constant aggregate initialization at file scope in a compound literal, but since that behavior will change soon if we implement support for dynamic initialization, I also added a unit test for `isConstantInitializer`.

This fixes at least one part of issue #80510 .

### CODE REVIEW

File: clang/test/SemaCXX/compound-literal.cpp [hunk 1]
```diff
@@ -108,3 +109,22 @@ int computed_with_lambda = [] {
...
+struct Empty {};
+struct Foo : Empty {
+  int x;
+  int y;
+};
+int f();
+Foo o = (Foo){
+  {},
+  1,
+  f() // expected-error {{initializer element is not a compile-time constant}}
```
Comment (shafik) [inline clang/test/SemaCXX/compound-literal.cpp:127]: gcc accepts this: https://godbolt.org/z/enWrG56je

Why do we believe this should be rejected?

File: clang/lib/AST/Expr.cpp [hunk 2]
```diff
@@ -3342,6 +3342,18 @@ bool Expr::isConstantInitializer(ASTContext &Ctx, bool IsForRef,
     if (ILE->getType()->isRecordType()) {
       unsigned ElementNo = 0;
       RecordDecl *RD = ILE->getType()->castAs<RecordType>()->getDecl();
+
+      // Check bases for C++17 aggregate initializers.
```
Comment (shafik) [inline clang/lib/AST/Expr.cpp:3346]: Should we be checking `CPlusPlus17` here? 

Also I would like to see a standard quote in the comments. I know we are not doing this locally but these are really helpful and we should be doing this more consistently.

**Review by shafik — COMMENTED** [review]
Also in the issue: https://github.com/llvm/llvm-project/issues/80510 you pointed out a crash bug. We should have the bug covered in the test case as well.

File: clang/test/SemaCXX/compound-literal.cpp (hunk 1 above)
Comment (rnk) [reply clang/test/SemaCXX/compound-literal.cpp:127]: That is correct and I did discover that in my testing, but I think fixing that goes beyond the scope of this patch. Clang currently rejects this case if you remove the empty base: https://godbolt.org/z/1x8hshM75 Adding support for non-constant compound literals at file scope requires codegen changes, and this is fixing a small bug to make more `isConstantInitializer` correct.

The real issue is that I couldn't find a better way to test `isConstantInitializer` that will be resilient to that future bug fix, and I'm not sure what to do about that.

Comment (rnk) [reply clang/test/SemaCXX/compound-literal.cpp]: I went ahead and make a targeted unit test for `isConstantInitializer` to see if that is nicer

**Comment by rnk** [comment]
> Also in the issue: #80510 you pointed out a crash bug. We should have the bug covered in the test case as well.

I think that is covered by the existing test case, instead of crashing, we now reject with an error. In the future if we want to align with GCC as you suggest, we'll need to have a codegen test case for non-constant compound literals of aggregates with bases.

**Comment by shafik** [comment]
@AaronBallman since this involves compound literals a C-ism, I would like you to review this as well.

File: clang/lib/AST/Expr.cpp [hunk 3]
```diff
@@ -3328,6 +3328,12 @@ bool Expr::isConstantInitializer(ASTContext &Ctx, bool IsForRef,
            DIUE->getUpdater()->isConstantInitializer(Ctx, false, Culprit);
   }
   case InitListExprClass: {
+    // C++ [temp.dep.expr]p2:
```
Comment (AaronBallman) [inline clang/lib/AST/Expr.cpp:3331]:
```suggestion
    // C++ [dcl.init.aggr]p2:
```
right?

**Review by AaronBallman — APPROVED** [review]
Please be sure to add a release note so users know about the fix, but aside from a bad standards reference, this LGTM (no concerns about C behavior either).

**Review by rnk — COMMENTED** [review]
Thanks! I added notes.

**Pull request merged**